qadence_protocols/measurements/shadow.py

```diff
@@ -4,90 +4,159 @@
-    observables: list[AbstractBlock],
-    options: dict,
-    param_values: dict[str, Tensor] = dict(),
-    state: Tensor | None = None,
-) -> Tensor:
-    """
-    Construct a classical shadow of a state to estimate observable expectation values.
-
-    Args:
-        model (QuantumModel): Model to evaluate.
-        observables (list[AbstractBlock]): a list of observables
-            to estimate the expectation values from.
-        param_values (dict): a dict of values to substitute the
-            symbolic parameters for.
-        options (dict): a dict of options for the measurement protocol.
-            Here, shadow_size (int), accuracy (float) and confidence (float) are supported.
-        state (Tensor | None): an initial input state.
-
-    Returns:
-        expectations (Tensor): an estimation of the expectation values.
-    """
-
-    circuit = model._circuit.original
-    shadow_size = options["shadow_size"]
-    accuracy = options["accuracy"]
-    confidence = options["confidence"]
-
-    return estimations(
-        circuit=circuit,
-        observables=observables,
-        param_values=model.embedding_fn(model._params, param_values),
-        shadow_size=shadow_size,
-        accuracy=accuracy,
-        confidence_or_groups=confidence,
-        state=state,
-        backend=model.backend,
-        noise=model._noise,
-        return_shadows=False,
-    )
+from qadence_protocols.measurements.abstract import MeasurementManager
+from qadence_protocols.measurements.utils_shadow import (
+    compute_snapshots,
+    expectation_estimations,
+    local_shadow,
+    number_of_samples,
+    shadow_samples,
+)
+from qadence_protocols.types import MeasurementData
+
+
+class ShadowManager(MeasurementManager):
+    """The class for managing randomized classical shadow."""
+
+    def __init__(self, measurement_data: MeasurementData = None, options: dict = dict()):
+        self.measurement_data = measurement_data
+        self.options = options
+
+    def verify_options(self) -> dict:
+        """Extract shadow_size, accuracy and confidence from options."""
+
+        shadow_size = self.options.get("shadow_size", None)
+        accuracy = self.options.get("accuracy", None)
+        if shadow_size is None and accuracy is None:
+            raise KeyError(
+                "Shadow protocol requires either an option"
+                " 'shadow_size' of type 'int' or 'accuracy' of type 'float'."
+            )
+        confidence = self.options.get("confidence", None)
+        if confidence is None:
+            raise KeyError("Shadow protocol requires an option 'confidence' of type 'float'.")
+        self.options = {"shadow_size": shadow_size, "accuracy": accuracy, "confidence": confidence}
+        return self.options
+
+    def number_of_samples_and_groups(
+        self, observables: list[AbstractBlock] = list()
+    ) -> tuple[int, ...]:
+        """Return the shadow size and number of groups for a list of observables.
```

Comment:
What is the number of groups here ?

Reply:
It is K, which is how many groups of measurements you use within the median of means algo part. Maybe I should call it differently. How about `n_medians` and `number_of_samples_and_medians`?

Reply:
Removed `number_of_samples_and_medians`